packages/convex-helpers/server/filterWith.ts

```diff
@@ -0,0 +1,216 @@
+/**
+ * new ReaderWithFilter(db)
+ * extends a DatabaseReader `db` to add a method `filterWith` method, which
+ * behaves the same as `filter` but allows arbitrary javascript filters.
+ */
+
+import {
+  DocumentByInfo,
+  GenericDataModel,
+  GenericDatabaseReader,
+  GenericTableInfo,
+  PaginationOptions,
+  QueryInitializer,
+  PaginationResult,
+  FilterBuilder,
+  Expression,
+  OrderedQuery,
+  IndexRange,
+  IndexRangeBuilder,
+  Indexes,
+  NamedIndex,
+  NamedSearchIndex,
+  Query,
+  SearchFilter,
+  SearchFilterBuilder,
+  SearchIndexes,
+  TableNamesInDataModel,
+  DocumentByName,
+  NamedTableInfo,
+} from "convex/server";
+import { GenericId } from "convex/values";
+
+async function asyncFilter<T>(
+  arr: T[],
+  predicate: (d: T) => Promise<boolean>
+): Promise<T[]> {
+  const results = await Promise.all(arr.map(predicate));
+  return arr.filter((_v, index) => results[index]);
+}
+
+type Predicate<T extends GenericTableInfo> = (
+  doc: DocumentByInfo<T>
+) => Promise<boolean>;
+
+export class QueryWithFilter<T extends GenericTableInfo> implements Query<T> {
+  q: Query<T>;
+  p: Predicate<T>;
+  iterator?: AsyncIterator<any>;
+
+  constructor(q: Query<T> | OrderedQuery<T>, p: Predicate<T>) {
+    this.q = q as Query<T>;
+    this.p = p;
+  }
+  filter(predicate: (q: FilterBuilder<T>) => Expression<boolean>): this {
+    return new QueryWithFilter(this.q.filter(predicate), this.p) as this;
+  }
+  filterWith(predicate: Predicate<T>): this {
+    return new QueryWithFilter(this.q, async (d) => {
+      return (await this.p(d)) && (await predicate(d));
+    }) as this;
+  }
+  order(order: "asc" | "desc"): QueryWithFilter<T> {
+    return new QueryWithFilter(this.q.order(order), this.p);
+  }
```

Comment:
this means you can call `.order` multiple times on a query

Reply:
yeah darn. you sort of can anyway in javascript because `QueryImpl` implements both Query and OrderedQuery, but it would be good for the types to not allow it